Basic health, first iteration

stilnat opened the pull request:
<!-- The notes within these arrows are for you but can be deleted. -->

<!-- Add "[WIP]" to the beginning of your title if you aren't immediately ready for review. -->

## Summary

This PR has for objective implementation of the first iteration of the health system, following this roadmap : 
https://github.com/RE-SS3D/SS3D/issues/1192

Most is said in the road map.

To test it :

- Put your mouse cursor on a body part of your own and press F, you should take some damages and eventually a limb is removed.
- Try using "hurt" and "examinebodypart" commands to inflict different kind of damages and check they were applied correctly.
- Try using the Hit interaction by clicking on another player and pressing hit, it should inflict some damages to the head, repeat and see them die.
- See that with no hands, you can't pick up stuff.
- See that you can't run with no feet (try hurting your feet).

## Pictures/Videos (optional)



https://github.com/RE-SS3D/SS3D/assets/14344825/63f1a999-fc81-4f36-bafc-1bb6dae2a3be


## Changes to Files


## Technical Notes (optional)

Bodyparts are networked objects, but the body layers they are made of are not, so their value are server only for now (and
it might be better to leave it that way). 
They can of course be queried for specific purpose, such is done to update player speed depending on body layers damages on
foot.

## Known issues



## Fixes (optional)

## TODO

- [x] clean and organise files
- [x] comment all the code
- [x] Hide brain on clients
- [x] fix indentation

stilnat committed:
add basic health files

stilnat committed:
fix kill interaction

stilnat committed:
add blood particles

stilnat committed:
add hand body part

stilnat committed:
fix back human prefab

stilnat committed:
update v3 fishnet

stilnat committed:
fixing icons not showing up

stilnat committed:
fix some warning and issues

stilnat committed:
Merge branch 'update-fishnet-v3' into basic-health-v3

stilnat committed:
remove require component network object on bodypart

stilnat committed:
allow attach body parts to be destroyed

stilnat committed:
add bodyparts on human

stilnat committed:
add bodyparts tag

stilnat committed:
allow to detach hands

stilnat committed:
spawn body part when destroyed

stilnat committed:
add hurt coimmand to hurt a player

stilnat committed:
remove name as a member of bodypart

stilnat committed:
move scripts on bodyparts for bodyparts instead of the armature

stilnat committed:
same

stilnat committed:
rename some clothes gameobject

stilnat committed:
move hand and hand container on bodyparts

stilnat committed:
rename correctly some gameobjects

stilnat committed:
remove prefab member of item

stilnat committed:
remove useless log

stilnat committed:
remove exception

stilnat committed:
add item prefabs for body parts

stilnat committed:
make head convex

stilnat committed:
replace prefab to spawn when bodypart destroyed

stilnat committed:
some factoring and destroy bodyparts instead of only detach

stilnat committed:
mode id container on body parts

stilnat committed:
move ears on head

stilnat committed:
tidy up human scripts and update doc

stilnat committed:
add head body part

stilnat committed:
fix some issues with removing body parts

stilnat committed:
fix null ref

stilnat committed:
change body parts collider to be triggers

stilnat committed:
can invoke disabled attached container

stilnat committed:
remove items from bodyparts before detaching

stilnat committed:
fix is kinematic on hand left item

stilnat committed:
some factoring on bodypart

stilnat committed:
update attack by pressing f

stilnat committed:
update default prefab objects for fiishnet

stilnat committed:
add entity on head when head detached

stilnat committed:
refactor some stuff

stilnat committed:
fix screen going black when non owned entity gets destroyed

stilnat committed:
simple script to check who owns a given go

stilnat committed:
fix ownership on clients when picking up stuff

stilnat committed:
add brain to head

stilnat committed:
kill human brain by destroying head

stilnat committed:
add ghost prefab

stilnat committed:
add two commands to hurt body parts

stilnat committed:
fix some stuff on bodyparts

stilnat committed:
set ref to ghost

stilnat committed:
refresh default prefab objects

stilnat committed:
update bodyparts to fix rigidbody interpolation

stilnat committed:
update human

stilnat committed:
copy value to bodypart

stilnat committed:
add ability to hit all layer of bodypart

stilnat committed:
add examine body part command

stilnat committed:
fix modified damagetypeQuantity

stilnat committed:
Merge branch 'develop' into basic-health-v3

stilnat committed:
fix isdetached check

stilnat committed:
fix detached bodypart

stilnat committed:
some doc and clean up

stilnat committed:
simplified detach body part workflow

stilnat committed:
removing useless interactive container on player

stilnat committed:
no interactions for now

stilnat committed:
fix head body part issue with detaching

stilnat committed:
add hit interaction, for a combat basis

stilnat committed:
some cleanup

stilnat committed:
destroy container content when body part is destroyed

stilnat committed:
add low speed with low foot health

stilnat committed:
remove serializers for body layers

stilnat committed:
move files around

stilnat committed:
add feet script to human

stilnat committed:
Hide brain

stilnat committed:
add doc

stilnat committed:
replace tabs with 4 spaces

stilnat committed:
hide body on head severed

stilnat committed:
add cloth container script

stilnat committed:
add organ container

avoid dropping organs in body part containers on detached

joaoburatto commented on Assets/Scripts/SS3D/Systems/Health/DamageType.cs:
```diff
@@ -0,0 +1,14 @@
+    Rad,
+    Toxic,
+    Acid,
+    Oxy,
```
"Respiratory" maybe? I don't feel its correct to use any simplifications like "Oxy", could lead to future confusion

cosmiccoincidence replied:
Oxy is what ss13 uses for basically the same thing so I dont think there will be much confusion there.

But there could be confusion calling it respiratory because of the respiratory system (body layer) the character has.

stilnat committed:
correct namespaces

stilnat committed:
change oxy to respiratory

stilnat committed:
add human entity and some fix

got some issues ... when head is cut, a head item spawn and there's a mind swap on the new entity which works okay. However this "new" entity (the head), doesn't register in the entity system and is not found by it. I wanted to simply add it to the _spawnedPlayers synclist but it triggers a bunch of other stuff in the game mode system and such and I'm unsure if it's any wanted behavior ... fixed by setting properly the synclist in entitySystem

stilnat committed:
separating feetcontroller and humanoid one

stilnat committed:
remove inventory UI when inventory is disabled.

stilnat committed:
add possibility of hurting all body layers in hurt command

stilnat committed:
update to develop

stilnat committed:
Revert "change oxy to respiratory"

This reverts commit e9633df5bccc2321b40f5e5a39caf1b09d11177e.
